Restructure library to conform to modern standards and best-practices

cbj4074 opened the pull request:
This entire library should be rewritten, conforming to modern standards (e.g., PSRs) and best-practices, with accompanying tests.

This PR serves as a starting-point, and aims to "get us to a good place", so we're in a position to articulate the requirements through tests and then rework the code-base until they're all passing.

cbj4074 committed:
docs: License under MIT

cbj4074 committed:
chore: Update dependencies

cbj4074 committed:
wip: Initial substantive commit for complete rewrite

cbj4074 committed:
wip: Flesh-out the Process component a bit further

cbj4074 committed:
tests: Add more realistic test for FLAC-to-MP3 encoding

cbj4074 committed:
feat: Add simple Utility class whose contents may be moved elsewhere

cbj4074 committed:
refact: Begin reworking transcode() method

cbj4074 committed:
style: Fix formatting

poppabear8883 commented on src/ProcessFailedException.php:
```diff
@@ -0,0 +1,13 @@
+
+class ProcessFailedException extends SymfonyProcessFailedException
+{
+    public function __construct(Process $process)
```
Might be a good idea to code to the `ProcessInterface` here ?

cbj4074 replied:
Good catch; fixed in a subsequent commit.

cbj4074 committed:
feat: Add logging capabilities

cbj4074 committed:
refact: Move all DI/container logic into dedicated class

As part of this effort, a basic Bootstrapper has been added, too.

cbj4074 committed:
fix: Add support for environment variables to process-running

cbj4074 committed:
ci: Add PHP quality tools to dev-dependencies

cbj4074 committed:
refact: Type-hint the interface instead of the implementation

cbj4074 committed:
tests: Add tests to define additional requirements

cbj4074 committed:
tests: Correct invalid assertion

cbj4074 committed:
fix: Fix any runtime errors triggered by newly-added tests

Note: Ran php-cs-fixer, too.

cbj4074 committed:
style: Apply code quality tools

cbj4074 committed:
Fix: Restructure dependencies and add logging support to Tagger

cbj4074 committed:
feat: Add method to access STDERR output from Process

cbj4074 committed:
tests: Fix fact that WAV file does not exist

It's necessary to convert the FLAC file to a WAV file first.

cbj4074 committed:
Refact: Make tests pass

cbj4074 committed:
tests: Add tests for two exception scenarios

Also, change requirement for return types in a couple of instances.

cbj4074 committed:
refact: Eliminate unnecessary arguments in transcode() and pass tests

cbj4074 committed:
refact: Move Processing components into dedicated namespace

cbj4074 committed:
tests: Add test for audio trimming functionality

cbj4074 committed:
typo: Fix typo in comment

cbj4074 committed:
tests: Add docblocks and correct requirements in one test

cbj4074 committed:
feat: Add method to convert from HMS to seconds

cbj4074 committed:
docs: Add/fix docblocks

poppabear8883 commented on src/Transcoding/Transcoder.php:
```diff
@@ -39,19 +43,19 @@ public function __construct(
+        float $clipLength = 90,
+        float $fadeInLength = 0,
+        float $fadeOutLength = 0
     ) {
```
If opting to use `parameter types`, maybe its note worthy to mention `return types` ?

cbj4074 replied:
Agreed. I've added them in a couple of places, and think it makes sense to use them consistently throughout. This is our chance to modernize the library, and there's no incentive not to take advantage of this construct.

cbj4074 committed:
chore: Add lock file to source-control

poppabear8883 commented on tests/TranscodingTest.php:
```diff
@@ -0,0 +1,128 @@
+     */
+    public function testTranscodingFlacToMp3Succeeds()
+    {
+        $this->testDir =
```
what is this ??

cbj4074 replied:
An oversight. Oops! It should simply be removed.

poppabear8883 reviewed (changes requested):
Remove `Line #30` of `test/TranscodingTest.php`

cbj4074 committed:
wip: Completely re-architect the library

cbj4074 committed:
tests: Update test to match evolving requirements

cbj4074 committed:
fix: Add interfaces for the ManipulatorCreator classes

cbj4074 committed:
tests: Fix issues with FLAC conversion tests to MP3 and ALAC

cbj4074 committed:
feat: Add support for FLAC-to-ALAC conversion and pass test

cbj4074 committed:
tests: Add FLAC-specific test class

cbj4074 committed:
feat: Add FLAC-to-WAV conversion

cbj4074 committed:
tests: Move test method from defunct test class into new equivalent

cbj4074 committed:
tests: Move test methods from defunct class into new equivalent

cbj4074 committed:
feat: Add support for WAV conversion to several other formats

cbj4074 committed:
todo: Add a couple TODOs

cbj4074 committed:
tests: Restructure tests a bit

cbj4074 committed:
feat: Rough-out FLAC-tagging

And stage-up a couple of small MP3-tagging changes in the process.

poppabear8883 reviewed (changes requested):
I would like to see a SoxCommand builder object before this gets merged in :wink:

cbj4074 committed:
tests: Replace temporary approach with more realistic equivalent

cbj4074 committed:
fix: Make FLAC-tagging test pass

cbj4074 committed:
tests: Fix and improve several aspects of existing tests

cbj4074 committed:
refact: Remove cruft and replace return-false with exceptions

cbj4074 committed:
fix: Add missing exception class

cbj4074 committed:
refact: Delete defunct/unused code

cbj4074 committed:
fix: Fix incorrect namespaces for PSR-4 autoloading

cbj4074 committed:
tests: Fix output file locations

In the absence of a fully-qualified path, the files were being output
to a relative (and therefore fairly arbitrary) location.

cbj4074 committed:
tests: Add tests for CLI commands

cbj4074 committed:
refact: Improve CLI interaction

This still isn't ideal; another layer of abstraction is necessary.
Rather than type-hint the SoxCommandInterface on the constructor,
the EffectInterface should be used.

cbj4074 committed:
refact: Rework how audio converter effects are applied via CLI tools

poppabear8883 commented on src/CliCommand/SoxCommand.php:
```diff
@@ -0,0 +1,76 @@
+        return $this->name;
+    }
+
+    public function addPart(string $name, string $value): void
```
So seeing your approach here I would recommend refactoring the method name from `addPart()` .. I'm not real sure what to name it but you're not really adding the "part" but rather attaching to a part in the parts array ?

cbj4074 replied:
Good call; you're absolutely correct. What about `addArgumentToPart()`, or even just `addArgument()`?

poppabear8883 replied:
addArgument(); sounds good to me ?

poppabear8883 commented on src/CliCommand/SoxCommand.php:
```diff
@@ -0,0 +1,76 @@
+
+    protected $binary = 'sox';
+
+    protected $parts = [
```
Make this an over-writable property in the base class

cbj4074 replied:
What should its initial value be? An empty array, e.g., `protected $parts = [];`?

poppabear8883 replied:
make it an abstract property (forcing the implementation to overwrite it) and yes an empty array

poppabear8883 replied:
ignore that 😕

poppabear8883 replied:
yes initialize it to an empty array or `null` ? if `null` then we can `test` if its been set in the implementation, if not through an exception

poppabear8883 commented on src/CliCommand/SoxCommand.php:
```diff
@@ -0,0 +1,76 @@
+{
+    private $name = 'sox_command';
+
+    protected $binary = 'sox';
```
Make this an over-writable property in the base class

cbj4074 replied:
Same question here; should the value be initialized to an empty string? `null`?

poppabear8883 commented on src/CliCommand/SoxCommand.php:
```diff
@@ -0,0 +1,76 @@
+
+class SoxCommand extends CliCommand implements SoxCommandInterface
+{
+    private $name = 'sox_command';
```
Make this an over-writable property in the base class

poppabear8883 commented on src/CliCommand/SoxCommandInterface.php:
```diff
@@ -0,0 +1,10 @@
+
+interface SoxCommandInterface extends CliCommandInterface
+{
+
```
Add your methods here :wink:

cbj4074 replied:
Oops! Thanks. :)

poppabear8883 reviewed (commented):
Couple comments, Overall the approach seems to be on point with what was discussed, some fine tuning and testing should reveal what needs to be accounted for.

cbj4074 committed:
refact: Fix a few missing properties and improve a method name

cbj4074 committed:
tests: Rework audio test sample files and directory structure

cbj4074 committed:
tests: Modify setUp() bodies to copy sample files to tmp directory

Absent this measure, the chance of introducing side-effects
(by modifying the sample file[s] inadvertently) is much greater.

cbj4074 committed:
tests: Satisfy the need to have a WAV sample

Rather than store sample files in every format, an alternative is to
store a single FLAC file that can be converted to any other format that
could ever be needed. This commit demonstrates one approach.

cbj4074 committed:
style: Remove superfluous root namespace indicator

cbj4074 committed:
tests: Rename CliCommandTest to SoxCommandTest and improve method

cbj4074 committed:
tests: Add MP3 tagging test

cbj4074 committed:
feat: Implement MP3-tagging logic required to make tests pass

cbj4074 committed:
fix: Add mechanism to ensure that required properties are set

cbj4074 committed:
tests: Delete unnecessary test

cbj4074 committed:
tests: Further-refine SoX command requirements

cbj4074 committed:
feat: Add fade() method and make all tests pass

cbj4074 committed:
fix: Add a couple keywords and delete lock file as it's not needed

cbj4074 committed:
chore: Reduce required phpunit version

Support for PHP 7.4 is as yet unnecessary, whereas 7.1 is a must.

cbj4074 committed:
ci: Add required CLI tools to Travis-CI build and add badges to README

cbj4074 commented:
> 
> 
> I would like to see a SoxCommand builder object before this gets merged in 😉

I've implemented that. I'm not 100% happy with it yet (I'll explain why in a subsequent note), but it's a start.

cbj4074 committed:
ci: Modify dependencies to suit Ubuntu 16

cbj4074 committed:
feat: Make key binary paths configurable per-environment

cbj4074 committed:
ci: Set an appropriate value for the ffmpeg binary on Ubuntu 16

cbj4074 committed:
fix: Fix string-to-array conversion issue

Shame on me for not running tests prior to committing!

cbj4074 committed:
fix: Use getenv() function instead of $_ENV superglobal
